### DISCUSSION TITLE
Convert coordinate of vehicles in 3D to the image coordinate

### DISCUSSION THREAD

**Discussion by hailuu684** [carla-simulator/carla#5229] [Q&A] [answered]
Hello everyone,
I am trying to convert 3D coordinate of vehicles in carla to image coordinate using pinhole equation, however, I do not know if I did it correctly. 
This is my code 
`def find_u_v(FOV, imgSize_x, imgSize_y, location):`
   ```
 Px = imgSize_x/2
    Py = imgSize_y/2
    focal_length = imgSize_x / (2 * np.tan(FOV * np.pi / 360))
    x = location.x
    y = location.y
    z = location.z
    u = focal_length * (-y) / x + Px
    v = focal_length * (-z) / x + Py
    return u, v
```

because the pinhole equation has a different axis orientation, in detail, x is right, y is up, z is forward. and the axis of carla is x-forward, y-right, z-up. Therefore, I just simply changed the values, I mean I consider x-axis in carla as z-axis in pinhole equation. However, I think it is not correct and I do not know I was wrong.

**Comment by LieYuan1993** [comment]
Hi, did you solve this problem? I have the same problem. I guess it is because the Carla coordinate system is a little different.

**Reply by hailuu684** [reply]
Hi, no actually. I still cannot find any solution for this

**Reply by AOOOOOA** [reply]
> Hi, did you solve this problem? I have the same problem. I guess it is because the Carla coordinate system is a little different.

I agree. When using the normal transformation in Carla, the result has some deviations.

**Comment by KevinLADLee** [comment, marked as answer]
Hi @hailuu684 , I have a dataset tool to label the data collected form carla to KITTI-Object format, which exactly what you want to do. 

FYI:
https://github.com/KevinLADLee/carla_dataset_tools/blob/master/label_tools/kitti_objects_label.py

**Reply by hailuu684** [reply]
Hello, thank you for the help

**Comment by hailuu684** [comment]
Hello guys,
This code can help to solve the problem

```
def relative_coordinate_1(target_data, camera_rgb, K):
    # import transforms3d
    # import math
    """
    camera rgb of ego vehicle, also remember to change camera_rgb and K matrix according to left, right and back camera
    """
    # This (4, 4) matrix transforms the points from world to sensor coordinates.
    world_2_camera = np.array(camera_rgb.get_transform().get_inverse_matrix())
    world_points = np.array([target_data.get_location().x, target_data.get_location().y, target_data.get_location().z, 1]).T

    # Transform the points from world space to camera space.
    sensor_points = np.dot(world_2_camera, world_points)
    point_in_camera_coords = np.array([sensor_points[1], sensor_points[2] * -1, sensor_points[0]]).T

    # Finally we can use our K matrix to do the actual 3D -> 2D.
    points_2d = np.dot(K, point_in_camera_coords)

    # Remember to normalize the x, y values by the 3rd value.
    points_2d = np.array([points_2d[0] / points_2d[2], points_2d[1] / points_2d[2], points_2d[2]])

    return points_2d[0], points_2d[1] - 10
```
